utils/service_discovery/sd_docker_backend.py

```diff
@@ -47,31 +91,42 @@ def __init__(self, agentConfig):
 
         AbstractSDBackend.__init__(self, agentConfig)
 
+    def _make_fetch_state(self):
+        return _SDDockerBackendConfigFetchState(
+            self.docker_client.inspect_container,
+            self.kubeutil.retrieve_pods_list().get('items', []) if Platform.is_k8s() else None)
+
     def update_checks(self, changed_containers):
-        conf_reload_set = set()
-        for id_ in changed_containers:
-            try:
-                inspect = self.docker_client.inspect_container(id_)
-            except (NullResource, NotFound):
-                inspect = {}
+        state = self._make_fetch_state()
 
-            checks = self._get_checks_from_inspect(inspect)
+        conf_reload_set = set()
+        for c_id in changed_containers:
+            checks = self._get_checks_to_refresh(state, c_id)
             conf_reload_set.update(set(checks))
 
         if conf_reload_set:
             self.reload_check_configs = conf_reload_set
 
-    def _get_checks_from_inspect(self, inspect):
+    def _get_checks_to_refresh(self, state, c_id):
         """Get the list of checks applied to a container from the identifier_to_checks cache in the config store.
         Use the DATADOG_ID label or the image."""
+        inspect = state.inspect_container(c_id)
```

Comment:
So we actually have another issue to take care of here, which is that if the container which the ID is from is already removed at this point, inspect will be None and we won't reload anything although there's a check that now fails on a missing container.

To be clear it's a pre-existing bug, not yours, but let's fix it here. In this situation we can't really decide which check to reload since the container isn't here anymore, so I think we should trigger a full reload.
Adding this block here does the job:
```python
        # if the container was removed we can't tell which check is concerned
        # so we have to reload everything
        if not inspect:
            self.reload_check_configs = True
            return
```